File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/model/SearchParamsTest.java
package com.dbottillo.mtgsearchfree.model;

import android.os.Parcel;
import android.test.suitebuilder.annotation.SmallTest;

import com.dbottillo.mtgsearchfree.BaseTest;

import org.junit.Before;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

@SmallTest
public class SearchParamsTest extends BaseTest {

    SearchParams searchParams;

    @Before
    public void setup() {
        searchParams = new SearchParams();
        searchParams.setName("jayce");
        searchParams.setTypes("creatures");
        searchParams.setText("text");
        searchParams.setCmc(new IntParam("=", 4));
        searchParams.setPower(new IntParam(">", 3));
        searchParams.setTough(new IntParam("<", 1));
        searchParams.setWhite(false);
        searchParams.setBlack(false);
        searchParams.setBlue(true);
        searchParams.setRed(true);
        searchParams.setGreen(false);
        searchParams.setOnlyMulti(true);
        searchParams.setCommon(true);
        searchParams.setUncommon(true);
        searchParams.setRare(false);
        searchParams.setMythic(false);
        searchParams.setSetId(3);
    }

    @Test
    public void searchParams_ParcelableWriteRead() {
        Parcel parcel = Parcel.obtain();
        searchParams.writeToParcel(parcel, searchParams.describeContents());
        parcel.setDataPosition(0);

        SearchParams createdFromParcel = SearchParams.CREATOR.createFromParcel(parcel);
        assertThat(createdFromParcel, is(searchParams));
    }

    @Test
    public void searchParams_willDetectOneColor() {
        assertTrue(searchParams.atLeastOneColor());
        searchParams.setBlue(false);
        searchParams.setRed(false);
        assertFalse(searchParams.atLeastOneColor());
    }

    @Test
    public void searchParams_willDetectOneRarity() {
        assertTrue(searchParams.atLeastOneRarity());
        searchParams.setCommon(false);
        searchParams.setUncommon(false);
        assertFalse(searchParams.atLeastOneRarity());
    }

    @Test
    public void searchParams_validation() {
        SearchParams searchParams = new SearchParams();
        assertFalse(searchParams.isValid());

        searchParams = new SearchParams();
        searchParams.setText("text");
        assertTrue(searchParams.isValid());

        searchParams = new SearchParams();
        searchParams.setTypes("types");
        assertTrue(searchParams.isValid());

        searchParams = new SearchParams();
        searchParams.setName("search");
        assertTrue(searchParams.isValid());

        searchParams = new SearchParams();
        searchParams.setCmc(new IntParam("=", 4));
        assertTrue(searchParams.isValid());

        searchParams = new SearchParams();
        searchParams.setPower(new IntParam("=", 4));
        assertTrue(searchParams.isValid());

        searchParams = new SearchParams();
        searchParams.setTough(new IntParam("=", 4));
        assertTrue(searchParams.isValid());

        searchParams = new SearchParams();
        searchParams.setGreen(true);
        assertTrue(searchParams.isValid());

        searchParams = new SearchParams();
        searchParams.setRare(true);
        assertTrue(searchParams.isValid());

        searchParams = new SearchParams();
        searchParams.setSetId(2);
        assertTrue(searchParams.isValid());
    }

    @Test
    public void searchParams_willHandleMulticolorOption() {
        SearchParams searchParams = new SearchParams();
        assertFalse(searchParams.isNoMulti());
        assertFalse(searchParams.onlyMulti());
        searchParams.setNoMulti(true);
        assertTrue(searchParams.isNoMulti());
        assertFalse(searchParams.onlyMulti());
        searchParams.setOnlyMulti(true);
        assertFalse(searchParams.isNoMulti());
        assertTrue(searchParams.onlyMulti());
    }
}

File: MTGSearch/src/test/kotlin/com/dbottillo/mtgsearchfree/model/SearchParamsTest.kt
package com.dbottillo.mtgsearchfree.model

import org.hamcrest.CoreMatchers.`is`
import org.junit.Assert.*
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.junit.MockitoJUnit

class SearchParamsTest {

    @Rule
    @JvmField
    var mockitoRule = MockitoJUnit.rule()

    lateinit var searchParams: SearchParams

    @Before
    fun setup() {
        searchParams = SearchParams()
        searchParams.name = "jayce"
        searchParams.types = "creatures"
        searchParams.text = "text"
        searchParams.cmc = CMCParam("=", 4, listOf("4"))
        searchParams.power = PTParam(">", 3)
        searchParams.tough = PTParam("<", 1)
        searchParams.isWhite = false
        searchParams.isBlack = false
        searchParams.isBlue = true
        searchParams.isRed = true
        searchParams.isGreen = false
        searchParams.setOnlyMulti(true)
        searchParams.isCommon = true
        searchParams.isUncommon = true
        searchParams.isRare = false
        searchParams.isMythic = false
        searchParams.setId = 3
    }

    @Test
    fun searchParams_willDetectOneColor() {
        assertTrue(searchParams.atLeastOneColor())
        searchParams.isBlue = false
        searchParams.isRed = false
        assertFalse(searchParams.atLeastOneColor())
    }

    @Test
    fun searchParams_willDetectOneRarity() {
        assertTrue(searchParams.atLeastOneRarity())
        searchParams.isCommon = false
        searchParams.isUncommon = false
        assertFalse(searchParams.atLeastOneRarity())
    }

    @Test
    fun searchParams_validation() {
        var searchParams = SearchParams()
        assertFalse(searchParams.isValid)

        searchParams = SearchParams()
        searchParams.text = "text"
        assertTrue(searchParams.isValid)

        searchParams = SearchParams()
        searchParams.types = "types"
        assertTrue(searchParams.isValid)

        searchParams = SearchParams()
        searchParams.name = "search"
        assertTrue(searchParams.isValid)

        searchParams = SearchParams()
        searchParams.cmc = CMCParam("=", 4, listOf("4"))
        assertTrue(searchParams.isValid)

        searchParams = SearchParams()
        searchParams.power = PTParam("=", 4)
        assertTrue(searchParams.isValid)

        searchParams = SearchParams()
        searchParams.tough = PTParam("=", 4)
        assertTrue(searchParams.isValid)

        searchParams = SearchParams()
        searchParams.isGreen = true
        assertTrue(searchParams.isValid)

        searchParams = SearchParams()
        searchParams.isRare = true
        assertTrue(searchParams.isValid)

        searchParams = SearchParams()
        searchParams.setId = 2
        assertTrue(searchParams.isValid)
    }

    @Test
    fun isNotValid() {
        val searchParams = SearchParams()
        assertThat(searchParams.isValid, `is`(false))

        searchParams.power = null
        assertThat(searchParams.isValid, `is`(false))
    }

    @Test
    fun searchParams_willHandleMulticolorOption() {
        val searchParams = SearchParams()
        assertFalse(searchParams.isNoMulti)
        assertFalse(searchParams.onlyMulti())
        searchParams.isNoMulti = true
        assertTrue(searchParams.isNoMulti)
        assertFalse(searchParams.onlyMulti())
        searchParams.setOnlyMulti(true)
        assertFalse(searchParams.isNoMulti)
        assertTrue(searchParams.onlyMulti())
    }

    @Test
    fun returnValidWithOnlyLandSet() {
        val searchParams = SearchParams()
        searchParams.isLand = true
        assertTrue(searchParams.isValid)
    }
}